aria/modeling/service.py

```diff
@@ -0,0 +1,1529 @@
+
+    @declared_attr
+    def groups(cls):
+        return cls.one_to_many_relationship('group')
+
+    @declared_attr
+    def policies(cls):
+        return cls.one_to_many_relationship('policy')
+
+    @declared_attr
+    def substitution(cls):
+        return cls.one_to_one_relationship('substitution')
+
+    @declared_attr
+    def inputs(cls):
+        return cls.many_to_many_relationship('parameter', table_prefix='inputs',
+                                             dict_key='name')
+
+    @declared_attr
+    def outputs(cls):
+        return cls.many_to_many_relationship('parameter', table_prefix='outputs',
+                                             dict_key='name')
+
+    @declared_attr
+    def operations(cls):
+        return cls.one_to_many_relationship('operation', dict_key='name')
+
+    @declared_attr
+    def plugins(cls):
+        return cls.many_to_many_relationship('plugin')
+
+    created_at = Column(DateTime, nullable=False, index=True)
+    updated_at = Column(DateTime)
+
+    # region orchestration
+
+    permalink = Column(Text)
+    scaling_groups = Column(modeling_types.Dict)
+
+    # endregion
+
+    # region foreign keys
+
+    __private_fields__ = ['substituion_fk',
+                          'service_template_fk']
+
+    # Service one-to-one to Substitution
+    @declared_attr
+    def substitution_fk(cls):
+        return cls.foreign_key('substitution', nullable=True)
+
+    # Service many-to-one to ServiceTemplate
+    @declared_attr
+    def service_template_fk(cls):
+        return cls.foreign_key('service_template', nullable=True)
+
+    # endregion
+
+    def satisfy_requirements(self, context):
+        satisfied = True
+        for node in self.nodes:
+            if not node.satisfy_requirements(context):
+                satisfied = False
+        return satisfied
+
+    def validate_capabilities(self, context):
+        satisfied = True
+        for node in self.nodes:
+            if not node.validate_capabilities(context):
+                satisfied = False
+        return satisfied
+
+    def find_nodes(self, node_template_name):
+        nodes = []
+        for node in self.nodes:
+            if node.node_template.name == node_template_name:
+                nodes.append(node)
+        return collections.FrozenList(nodes)
+
+    def get_node_ids(self, node_template_name):
+        return collections.FrozenList((node.name for node in self.find_nodes(node_template_name)))
+
+    def find_groups(self, group_template_name):
+        groups = []
+        for group in self.groups:
+            if group.template_name == group_template_name:
+                groups.append(group)
+        return collections.FrozenList(groups)
+
+    def get_group_ids(self, group_template_name):
+        return collections.FrozenList((group.name
+                                       for group in self.find_groups(group_template_name)))
+
+    def is_node_a_target(self, context, target_node):
```

Comment:
everynode which is a target has inbound_relationships. these methods are probably not relavent anymore